drivers/ethernet/eth_stm32_hal.c

```diff
@@ -1195,19 +1195,30 @@ static int ptp_clock_stm32_get(const struct device *dev,
 	struct eth_stm32_hal_dev_data *eth_dev_data = ptp_context->eth_dev_data;
 	ETH_HandleTypeDef *heth = &eth_dev_data->heth;
 	int key;
+	uint32_t second_2;
 
 	key = irq_lock();
 
 #if defined(CONFIG_SOC_SERIES_STM32H7X)
 	tm->second = heth->Instance->MACSTSR;
 	tm->nanosecond = heth->Instance->MACSTNR;
+	second_2 = heth->Instance->MACSTSR;
 #else
 	tm->second = heth->Instance->PTPTSHR;
 	tm->nanosecond = heth->Instance->PTPTSLR;
+	second_2 = heth->Instance->PTPTSHR;
 #endif /* CONFIG_SOC_SERIES_STM32H7X */
 
 	irq_unlock(key);
 
+	if (tm->second != second_2 && tm->nanosecond < NSEC_PER_SEC / 2) {
```

Comment:
Why do we need to check the nanosecond value here?

Reply:
It's possible that second_2 != tm->second (was read after rollover), but nanosecond is 999999999 (read before rollover). If that's true, we don't want to do adjustment. We only want to do adjustment if nanosecond rolled over and second did not.